### PR TITLE
Fix: Creatability of page types incorrectly affecting the `can_reorder_subpages()` check

### PR DESCRIPTION
**Pull request by ababic**
Fixes https://github.com/wagtail/wagtail/issues/3465

Where all of the page types in `subpage_types` have `is_creatable=False` on them, it is currently not possible to reorder pages that exist below instances of that page type.

I'm classifying this as a bug, because I have always interpreted the `is_creatable=False` mechanism as a way to prevent manual creation ONLY. Not only do I feel it is perfectly valid for programatically-created pages to exist at certain locations (and remain editable/publishable/deletable). But, projects also have a life outside of the current state of the codebase... it's quite possible for pages to have been created before the current code state (where `is_creatable` or `subpage_types` have changed); In which case, the ability to manage those pages (including 'reordering') should not be impacted.

### CODE REVIEW

**Review by laymonage — COMMENTED** [review]
Thanks @ababic!

I was also testing with `subpage_types = []` which would now show the "Sort menu order" option as well, even though you cannot actually reorder the children (since the operation would basically still do a "move" to the parent page). However, this is an edge case that doesn't seem worth considering because this only happens if the child pages were created prior to setting the `subpage_types = []`. If there's a child page despite the model has `subpage_types = []`, that sounds like an unrelated data integrity issue.

That said, could you also add a test case on the reordering view itself, to ensure that this works when the action is performed?

We have a similar test case that relates to the `parent_page_types` restriction here:

https://github.com/wagtail/wagtail/blob/8ddf472e93710a68ec4a44fbc853239c53d25c60/wagtail/admin/tests/pages/test_reorder_page.py#L114-L142

Would be nice to have the same for `subpage_types` 🙂

**Comment by laymonage** [comment]
Hi @ababic, are you still keen on picking this up? No worries if not, I'll try to get to it soon. Thanks!

**Comment by ababic** [comment]
@laymonage I should have some headspace for this next week, unless you get to it sooner?

**Comment by laymonage** [comment]
@ababic I likely won't get to this until the week after, so if you have time next week that'd be great!

**Comment by ababic** [comment]
New tests aren't behaving quite as expected
- The tests are hitting the 'set_page_position' URL, which reaches the [set_page_position](https://github.com/wagtail/wagtail/blob/stable/6.1.x/wagtail/admin/views/pages/ordering.py#L8) view
- The only thing that could cause those views to redirect is the raising of a `PermissionDenied` error, which could only be raised by [this line](https://github.com/wagtail/wagtail/blob/stable/6.1.x/wagtail/admin/views/pages/ordering.py#L12C28-L12C81)
- The tests are run as a superuser, so permissions shouldn't be a problem in the updated `PagePermissionTester.can_reorder_children()` method

Confused :(

File: wagtail/admin/tests/pages/test_reorder_page.py [hunk 1]
```diff
@@ -140,3 +178,75 @@ def setUp(self):
...
+
+
+class TestPageReorderWithNoSubPageTypes(TestPageReorder):
+    """
+    This TestCase is the same as the TestPageReorder class above, but uses a
+    NoSubpageTypesPage instance as the parent page instead of SimplePage.
+
+    This ensures that a parent page having an empty `subpage_types` list does
+    not prevent the reorder pre-existing or automatically generated pages
+    beneath it.
```
Comment (laymonage) [inline wagtail/admin/tests/pages/test_reorder_page.py:190]: @ababic Thanks, sorry I didn't get to this before the 6.2 feature freeze.

The `PermissionDenied` was raised by the `MovePageAction` class which is called when the `move()` is done:

https://github.com/wagtail/wagtail/blob/9f518a007d73930c04d5c7e810c1f475190c82f6/wagtail/admin/views/pages/ordering.py#L32-L42

https://github.com/wagtail/wagtail/blob/9f518a007d73930c04d5c7e810c1f475190c82f6/wagtail/models/__init__.py#L2371-L2376

https://github.com/wagtail/wagtail/blob/9f518a007d73930c04d5c7e810c1f475190c82f6/wagtail/actions/move_page.py#L28-L33

The reason is because `PagePermissionTester.can_move_to(destination)` checks for `Page.can_move_to(destination)`, which in turn also checks for `Page.can_exist_under(parent)` that returns `False` when the moved page type is not in the parent's `subpage_types`.

While I initially considered this as an edge case that's not worth considering (https://github.com/wagtail/wagtail/pull/11944#pullrequestreview-2063002247), after reading this test's docstring, I think this is worth fixing too.

I pushed a commit [`7196066` (#11944)](https://github.com/wagtail/wagtail/pull/11944/commits/71960666691f921fecf80651a317b8e68a1c9b95) that should fix this. Do you think that's a reasonable fix? If so, I'll rebase and clean up the PR to be merged for 6.3. Thanks!

Comment (ababic) [reply wagtail/admin/tests/pages/test_reorder_page.py:190]: Yes, short of making `can_exist_under()` an instance method and handling this there (which would no doubt introduce further problems), your change appears to do the job nicely! It is a shame to introduce multiple meanings of 'can exist under'.

Comment (ababic) [reply wagtail/admin/tests/pages/test_reorder_page.py:190]: Thinking some more, how would you feel about adding a separate `can_exist_under_page()` method, put it right next to `can_exist_under()` and use docstrings to outline the difference? This would make the API much clearer overall, I think.

Comment (laymonage) [reply wagtail/admin/tests/pages/test_reorder_page.py:190]: Yeah, I wanted to add it inside `can_exist_under()` but unfortunately it's a classmethod. I'm not sure about adding a `can_exist_under_page()`, from the naming perspective it reads like the `_page` is referring to the parent page, which wouldn't be any different to the existing classmethod that also takes the parent page instance as a parameter. If we were to add a new method, maybe `page_can_exist_under()` helps avoid that confusion – but honestly I don't even understand why we have both `Page.can_move_to()` and `PagePermissionTester.can_move_to()`, it seems like the former's logic should've been implemented into the latter.

I opted to put the logic into `PagePermissionTester.can_move_to()` instead, and opened a new rebased and cleaned PR here, so others can review: [`534c5e4` (#12156)](https://github.com/wagtail/wagtail/pull/12156/commits/534c5e4b437e50a6383da0468f44e6ab0c8a15a5)

**Comment by laymonage** [comment]
Superseded by #12156.
